### ISSUE TITLE
[Feature] TurboQuant NPU Kernel Integration

### ISSUE THREAD

**Issue by sunghajung6688** [hw-native-sys/pypto-serving#20]
# TurboQuant KV-Cache Compression

TurboQuant (TQ) is an online KV-cache quantization path for Qwen3-14B on
Ascend NPU (910B/C, 950). It uses **PolarQuant** — random rotation followed
by optimal scalar (Lloyd-Max) quantization — to compress the K and V caches
from BF16 (~16 bits/value) to **4 bits/value** while the model runs. The
method is adapted from [TurboQuant (ICLR 2026)](https://arxiv.org/abs/2504.19874).

TQ is an **opt-in branch alongside the default FP (BF16) path**. It is
enabled with `--tq`; without the flag, serving is byte-for-byte identical to
the non-TQ path. TQ decode returns raw logits, so device sampling and
embedding are disabled in TQ mode and the engine samples on the host.

**Key choice:** pure 4-bit PolarQuant is used. The paper's 1-bit QJL
residual correction (Algorithm 2) is excluded — at the same 4-bit rate,
4-bit PolarQuant alone beats 3-bit PolarQuant + 1-bit QJL, and QJL's
amplified variance turns into attention noise after softmax.

## Usage

Offline generation (`examples/model/qwen3_14b/npu_generate.py`):

```bash
python examples/model/qwen3_14b/npu_generate.py \
  --model-dir /path/to/Qwen3-14B \
  --platform a2a3 \
  --tq \
  --max-seq-len 512 \
  --max-new-tokens 128
```

Serving:

```bash
pypto-serving \
  --model /path/to/Qwen3-14B \
  --backend npu \
  --platform a2a3 \
  --tq \
  --max-model-len 512 \
  --port 8899
```

Drop `--tq` to run the unchanged FP path.

## Architecture

The FP path is untouched. Every TQ addition is gated behind
`RuntimeConfig.kv_quant_config` (`KvQuantConfig`) and `tq_mode`:

```
--tq  →  RuntimeConfig.kv_quant_config (KvQuantConfig | None)
      →  model_runner:   enabled ? allocate UINT8 + FP32-scale pools
                                  : BF16 pools              (unchanged)
      →  npu_executor:   tq_mode ? _compile_tq_kernels (26-param kernels)
                                : existing FP compile     (unchanged)
      →  npu_runner:     TQ-aware _kv_bytes_per_page + 26-element arg tuples
```

Serving-repo files:

| File | Role |
| --- | --- |
| `config/types.py` | `KvQuantConfig` frozen dataclass; `RuntimeConfig.kv_quant_config`. |
| `model/common/runner/model_runner.py` | TQ KV-cache allocation (nibble-packed UINT8 + FP32 scales). |
| `model/qwen/npu_executor.py` | `tq_mode`; `supports_device_* = not tq_mode`; `_compile_tq_kernels` (rotation matrix + codebook, 26-param compile, shape validation). |
| `model/qwen/npu_runner.py` | TQ kernel-arg tuples; `_kv_bytes_per_page`; rot/codebook static uploads. |
| `model/qwen/qwen3_l3_dispatch.py` | `prefill_fwd_tq` / `decode_fwd_tq` bindings + host wrappers. |
| `cli/main.py`, `examples/model/qwen3_14b/npu_generate.py` | `--tq` / `--tq-mode` flag. |

The quantization kernels are vendored from pypto-lib under
`pypto-lib/models/qwen3/14b/` and loaded at compile time:

| File | Role |
| --- | --- |
| `turboquant_kv.py` | Lloyd-Max codebook; shared quantize/dequant primitives. |
| `qwen3_14b_prefill_tq_draft.py` | Full prefill forward with inline TQ compression. |
| `qwen3_14b_decode_tq_draft.py` | Decode forward with TQ dequant attention. |

Per-layer pipeline (prefill and decode share the same shape):

```
RMSNorm → Q/K/V proj → per-head Q/K RMSNorm
K/V: RoPE (K only) → L2 norm + normalize → rotate (R) → Lloyd-Max quantize
                                                   → store UINT8 idx + FP32 scale
Q:   per-head RMSNorm → RoPE
Attention: dequant K/V from cache (gather codebook × scale, unrotate) →
           QK matmul → softmax → SV matmul
out_proj + residual → RMSNorm → SwiGLU MLP + residual
```

### Core idea

Multiplying a vector by a random Haar-orthogonal matrix `R` spreads its energy
evenly: once `x̂` is unit-norm, each coordinate of `y = x̂ @ R` is
(asymptotically) i.i.d. `N(0, 1/d)`. Independent, identically distributed
coordinates are exactly the case a 1-D scalar quantizer handles optimally —
and for a Gaussian source the optimal scalar quantizer is the Lloyd-Max
codebook. So PolarQuant reduces vector quantization to: pull off the L2 norm,
rotate, and independently scalar-quantize every coordinate with one shared
Gaussian-optimal codebook.

Two consequences drive the design:

- The norm `γ = ‖x‖` is stored in FP32, so quantization error only perturbs the
  *direction* of the vector, never its magnitude. Attention scores `QKᵀ` are
  inner products, so low angular error preserves them.
- Because every coordinate shares one distribution, a single fixed 16-level
  codebook serves all dimensions, heads, and layers — no per-channel fitting.

V is consumed by a `softmax · V` weighted sum, where low mean-squared
reconstruction error is the right target; the same MSE-optimal PolarQuant
reconstruction serves it. This implementation applies pure 4-bit PolarQuant to
both K and V.

### Quantize

```
x ∈ R^d                      (post-RoPE for K, raw for V)
γ = sqrt(‖x‖² + ε)           L2 norm
x̂ = x / γ                    normalize
y = x̂ @ R                    random rotation (Haar orthogonal, BF16)
idx[j] = Σ 𝟙(y[j] ≥ b_i)    15 boundary comparisons → index ∈ [0, 15]
store (idx as nibble-packed UINT8, γ as FP32 scale)
```

### Dequantize

```
ŷ = codebook[idx]            FP32 centroids in rotated domain
ŷ = ŷ / ‖ŷ‖                  renormalize (correct quantization norm drift)
ŷ = ŷ · γ                    restore magnitude
x̂ = ŷ @ R^T                  inverse rotation back to original space
```

### Lloyd-Max codebook

After rotation, coordinates are i.i.d. `N(0, 1/d)`. Optimal 4-bit centroids
(16 levels) are solved analytically via Gaussian conditional expectation at
module load (`turboquant_kv.solve_lloyd_max()`); the 15 quantization
boundaries are the midpoints between adjacent centroids.

### Rotation matrix

- Per-layer dense orthogonal matrix (128×128, BF16), stacked as
  `[num_layers * head_dim, head_dim]`.
- Built via QR of a Gaussian matrix with consistent column signs:
  `Q = qr(randn)[0] * sign(diag(R))`.
- Seed `42 + layer * 1000` per layer (deterministic; must match between the
  CPU reference and the NPU).

### Scope of this implementation

The full TurboQuant paper offers more than is used here:

- **QJL residual (Algorithm 2):** (b−1)-bit PolarQuant plus 1-bit QJL for
  unbiased inner products. Excluded — see the top of this doc.
- **Fast rotation:** `D₁ H D₂` (random ±1 diagonals + Walsh-Hadamard),
  `O(d log d)` instead of `O(d²)`. Not used; a per-layer dense QR matrix is
  applied instead (cheap relative to the attention it feeds).
- **Mixed bit-rates:** outlier channels at higher bits (e.g. 2.5/3.5-bit).
  Not used; a flat 4-bit rate is applied to all coordinates.

## Cache layout

Per token, per KV head, one row for K and one for V:

| Field | Dtype | Bytes |
| --- | --- | --- |
| Nibble-packed indices (`head_dim // 2`) | UINT8 | 64 |
| L2 scale | FP32 | 4 |
| **Total per row** | | **68** |

Versus 256 bytes for a BF16 row (`head_dim × 2`), this is **~3.76×** per-row
compression. Allocation in `model_runner.init_kv_cache`:

```python
quant_shape = (cache_rows, config.head_dim // 2)   # UINT8, nibble-packed
scale_shape = (cache_rows, 1)                       # FP32
# quant_k/v_pages (UINT8) + k/v_scales_pages (FP32): four tensors
```

where `cache_rows = num_layers * num_pages * num_kv_heads * page_size`. Bytes
per page (all layers, K + V) is
`rows_per_page * 2 * (head_dim // 2 + 4)` (`npu_runner._kv_bytes_per_page`).
TQ requires `page_size = 128`, enforced by `_validate_supported_shape`.

## Constants

| Constant | Value | Meaning |
| --- | --- | --- |
| `N_LEVELS` | 16 | 4-bit quantization levels |
| `HEAD_DIM` | 128 | Attention head dimension |
| `NUM_KV_HEADS` | 8 | KV heads (GQA) |
| `NUM_HEADS` | 40 | Query heads |
| `BLOCK_SIZE` / `page_size` | 128 | KV cache page size |
| `TOK_TILE` | 16 | Prefill token block |
| `SEQ_TILE` | 128 | Prefill attention sequence tile |
| `CMP_TILE` | 64 | Fused dequant tile (A2A3 memory limit) |
| `CMP_CHUNK` | 32 | Dequant gather sub-tile (32-byte aligned) |
| `SB_BATCH` | 128 | Sequence-block batch for parallel attention |
| `Q_HEAD_BATCH` | 5 | Decode Q heads per attention group |
| `Q_HEAD_PAD` | 16 | Padded Q rows (cube alignment) |

## Attention dequant path

Both prefill and decode dequantize K/V inline in BF16 to match the FP path:

```
UINT8 idx → FP16 → INT32 → gather(codebook) → FP32 centroids
         → cast BF16 → rsqrt renorm → row_expand_mul(scale) → BF16
         → matmul(R^T) → FP32 → cast BF16       (unrotate)
         → copy to BF16 GM buffer → QK matmul (BF16 @ BF16^T)
```

K is unrotated via `ctx @ R^T` after attention; Q is **not** pre-rotated
(the Q pre-rotation block is disabled), matching the pure-PolarQuant strategy.

## Known issues & fixes

**Rotation matrix seed mismatch (NPU vs CPU).** The NPU used
`torch.manual_seed(42)` sequentially; the CPU reference used
`42 + layer * 1000` per layer with independent generators and normalized
`Q * sign(diag(R))`. Layer 0 matched, the rest diverged. Fixed by matching
the CPU exactly on both sides; after the fix prefill `quant_k` matches 99.3%
across all 40 layers and the rotation matrices are identical (max diff 0).

**Ascend BF16 matmul layout mismatch.** `matmul(q_padded, k_dequant^T)`
returned garbage even though Q and K each matched CPU individually. The cube
engine miscomputes when one input is a GM-slice BF16 tile and the other is a
register-format `cast(FP32→BF16)` tile. Fixed by copying the dequantized K
to a BF16 GM buffer in a separate scope, then slicing from it — both inputs
are now GM-slice BF16, matching the FP prefill pattern. Score self-consistency
went from cos ≈ 0 to ≈ 1.0.

**Decode Q RoPE batch-slice corruption.** Slicing a `[5, 128]` Q tile into a
`[5, 64]` half corrupts rows 0–1 (a pypto compiler bug), giving heads 0–1
scores of ~−1000 while heads 2–4 were correct. Fixed by processing each Q
head individually as `[1, HEAD_DIM]`, matching prefill. After the fix all Q
heads reach cos ≈ 1.0.

Two pypto kernel pitfalls to keep in mind: `pl.write` must capture its return
value or the write is dead-code-eliminated, and `pl.assemble` of a `[T, 1]`
ColMajor tensor only fills row 0 — flatten and write per-row.

## Validation

Accuracy validation against the FP / CPU reference is still in progress;
results will be filled in once testing completes.

Output quality (greedy):

> **FP:** "a Chinese multinational technology company that designs and sells
> consumer electronics, telecommunications equipment, and services. It is one
> of the world's largest and most valuable technology companies. Huawei was
> founded in 1987 by Ren Zhengfei…"
>
> **TQ:** "a Chinese multinational technology company headquartered in
> Shenzhen, Guangdong, China. It is the world's second-largest smartphone
> manufacturer by unit sales and the second-largest provider of
> telecommunications equipment…"

Both are factually correct and coherent; the wording difference is normal
sampling variation between independent runs.

## Performance

End-to-end TQ performance is currently **far below the FP path — under 1/3 of
FP throughput** — so TQ makes serving *slower* today, not faster. The expected
bandwidth win from the smaller KV cache has not shown up because the
quantize/dequant work added on every attention step dominates. This is a known
gap **pending operator-side (kernel) optimization** (fusing the
rotate/gather/unrotate dequant into the attention kernels), not a correctness
or algorithm problem: TQ is functionally correct but not yet
performance-competitive.

## Reference

- Paper: "TurboQuant: Online Vector Quantization with Near-optimal
  Distortion", ICLR 2026 — <https://arxiv.org/abs/2504.19874>.
